crates/sui-data-ingestion/src/workers/archival.rs

```diff
@@ -0,0 +1,189 @@
+            checkpoint_range: 0..0,
+            buffer: vec![],
+            summary_buffer: vec![],
+            last_commit_instant: Instant::now(),
+            should_update_progress: false,
+        };
+        Self {
+            remote_store,
+            state: Mutex::new(state),
+            commit_file_size: config.commit_file_size,
+            commit_duration: Duration::from_secs(config.commit_duration_seconds),
+        }
+    }
+
+    async fn upload(&self, state: &AccumulatedState) -> Result<()> {
+        let checkpoint_file_path =
+            format!("epoch_{}/{}.chk", state.epoch, state.checkpoint_range.start);
+        self.upload_file(
+            Path::from(checkpoint_file_path.clone()),
+            CHECKPOINT_FILE_MAGIC,
+            &state.buffer,
+        )
+        .await?;
+        let summary_file_path =
+            format!("epoch_{}/{}.sum", state.epoch, state.checkpoint_range.start);
+        self.upload_file(
+            Path::from(summary_file_path.clone()),
+            SUMMARY_FILE_MAGIC,
+            &state.summary_buffer,
+        )
+        .await?;
+        let manifest_bytes = self
+            .remote_store
+            .get(&Path::from("MANIFEST"))
+            .await?
+            .bytes()
+            .await?;
+        let mut manifest = read_manifest_from_bytes(manifest_bytes.to_vec())?;
+        let checkpoint_file_metadata = create_file_metadata(
+            std::path::Path::new(&checkpoint_file_path),
+            FileType::CheckpointContent,
+            state.epoch,
+            state.checkpoint_range.clone(),
+        )?;
+        let summary_file_metadata = create_file_metadata(
+            std::path::Path::new(&summary_file_path),
+            FileType::CheckpointSummary,
+            state.epoch,
+            state.checkpoint_range.clone(),
+        )?;
+        manifest.update(
+            state.epoch,
+            state.checkpoint_range.start,
+            checkpoint_file_metadata,
+            summary_file_metadata,
+        );
+        let bytes = finalize_manifest(manifest)?;
+        self.remote_store
+            .put(&Path::from("MANIFEST"), bytes)
+            .await?;
+        Ok(())
+    }
+
+    async fn upload_file(&self, location: Path, magic: u32, content: &[u8]) -> Result<()> {
+        let mut buffer = vec![0; 4];
+        BigEndian::write_u32(&mut buffer, magic);
+        buffer.push(StorageFormat::Blob.into());
+        buffer.push(FileCompression::Zstd.into());
+        buffer.extend_from_slice(content);
+        let mut compressed_buffer = vec![];
+        let mut cursor = Cursor::new(buffer);
+        compress(&mut cursor, &mut compressed_buffer)?;
+        self.remote_store
+            .put(&location, Bytes::from(compressed_buffer))
+            .await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Worker for ArchivalWorker {
+    async fn process_checkpoint(&self, checkpoint: CheckpointData) -> Result<()> {
+        let mut state = self.state.lock().await;
```

Comment:
It would be great to add some asserts to check if this `checkpoint` is the next one from the previous one in `state`

Reply:
it's ensured by the framework itself. Otherwise each pipeline would have to do such check. Or do you have other concern here?
another thing is with the framework progress is stored in meta store(dynamodb atm), maybe we need to add a check that at start manifest and DDB do not contradict each other